src/sqlancer/feldera/gen/FelderaExpressionGenerator.java

```diff
@@ -0,0 +1,289 @@
+import sqlancer.common.schema.AbstractTables;
+import sqlancer.feldera.FelderaGlobalState;
+import sqlancer.feldera.FelderaSchema;
+import sqlancer.feldera.ast.*;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public final class FelderaExpressionGenerator
+        extends TypedExpressionGenerator<FelderaExpression, FelderaSchema.FelderaColumn, FelderaSchema.FelderaDataType>
+        implements
+        NoRECGenerator<FelderaSelect, FelderaJoin, FelderaExpression, FelderaSchema.FelderaTable, FelderaSchema.FelderaColumn> {
+
+    private List<FelderaSchema.FelderaTable> tables;
+    private final FelderaGlobalState globalState;
+    private final int maxDepth;
+
+    public FelderaExpressionGenerator(FelderaGlobalState globalState) {
+        this.globalState = globalState;
+        this.maxDepth = globalState.getOptions().getMaxExpressionDepth();
+    }
+
+    private enum BooleanExpression {
+        NOT, COMPARISON, IS_NULL, BETWEEN;
+    }
+
+    private FelderaExpression getBinaryComparison(int depth) {
+        FelderaSchema.FelderaDataType type = getRandomType();
+        FelderaExpression left = generateExpression(type, depth + 1);
+        FelderaExpression right = generateExpression(type, depth + 1);
+        return new FelderaBinaryComparisonOperation(left, right,
+                FelderaBinaryComparisonOperation.FelderaComparisonOperator.getRandom());
+    }
+
+    private FelderaExpression generateBooleanExpression(int depth) {
+        BooleanExpression exprType = Randomly.fromOptions(BooleanExpression.values());
+        FelderaExpression expr;
+
+        switch (exprType) {
+        case NOT:
+            return new FelderaUnaryPrefixOperation(generateExpression(FelderaSchema.FelderaDataType.BOOLEAN, depth + 1),
+                    FelderaUnaryPrefixOperation.FelderaUnaryPrefixOperator.NOT);
+        case COMPARISON:
+            return getBinaryComparison(depth);
+        case IS_NULL:
+            return new FelderaUnaryPostfixOperation(generateExpression(getRandomType(), depth + 1),
+                    Randomly.fromOptions(FelderaUnaryPostfixOperation.FelderaUnaryPostfixOperator.IS_NULL,
+                            FelderaUnaryPostfixOperation.FelderaUnaryPostfixOperator.IS_NOT_NULL));
+        case BETWEEN:
+            FelderaSchema.FelderaDataType type = getRandomType();
+            expr = generateExpression(type, depth + 1);
+            FelderaExpression left = generateExpression(type, depth + 1);
+            FelderaExpression right = generateExpression(type, depth + 1);
+            return new FelderaBetweenOperation(expr, left, right,
+                    FelderaBetweenOperation.FelderaBetweenOperatorType.getRandom());
+        default:
+            throw new AssertionError(exprType);
+        }
+    }
+
+    private FelderaExpression generateStringExpression(int depth) {
+        // TODO
```

Comment:
in general any of these can have a call to a function with a suitable return type.

Reply:
Yes, this is a todo.